File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapViewContainer.java
package com.github.quadflask.react.navermap;

import android.os.Bundle;
import android.util.Log;
import android.view.Choreographer;
import android.view.MotionEvent;
import android.view.View;
import android.widget.FrameLayout;

import androidx.annotation.NonNull;

import com.facebook.react.bridge.LifecycleEventListener;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.uimanager.ThemedReactContext;
import com.naver.maps.geometry.LatLng;
import com.naver.maps.geometry.LatLngBounds;
import com.naver.maps.map.NaverMap;
import com.naver.maps.map.NaverMapOptions;

public class RNNaverMapViewContainer extends FrameLayout implements RNNaverMapViewProps {
    private RNNaverMapView mapView;
    private Bundle instanceStateBundle = new Bundle();
    private boolean isAttachedToWindow = false;

    public RNNaverMapViewContainer(@NonNull ThemedReactContext themedReactContext, ReactApplicationContext appContext, NaverMapOptions naverMapOptions) {
        super(ReactUtil.getNonBuggyContext(themedReactContext, appContext));
        this.mapView = new RNNaverMapView(themedReactContext, appContext, naverMapOptions, instanceStateBundle);
        addView(mapView);
    }

    // https://github.com/facebook/react-native/issues/17968#issuecomment-457236577
    private void setupLayoutHack() {
        Choreographer.getInstance().postFrameCallback(new Choreographer.FrameCallback() {
            @Override
            public void doFrame(long frameTimeNanos) {
                manuallyLayoutChildren();
                getViewTreeObserver().dispatchOnGlobalLayout();
                if (isAttachedToWindow)
                    Choreographer.getInstance().postFrameCallbackDelayed(this, 500);
            }
        });
    }

    private void manuallyLayoutChildren() {
        for (int i = 0; i < getChildCount(); i++) {
            View child = getChildAt(i);
            child.measure(MeasureSpec.makeMeasureSpec(getMeasuredWidth(), MeasureSpec.EXACTLY),
                    MeasureSpec.makeMeasureSpec(getMeasuredHeight(), MeasureSpec.EXACTLY));
            child.layout(0, 0, child.getMeasuredWidth(), child.getMeasuredHeight());
        }
    }

    @Override
    public boolean dispatchTouchEvent(MotionEvent ev) {
        if (mapView != null && mapView.getMap().getUiSettings().isScrollGesturesEnabled()) {
            switch (ev.getAction()) {
                case MotionEvent.ACTION_DOWN:
                case MotionEvent.ACTION_UP:
                    getParent().requestDisallowInterceptTouchEvent(true);
                    break;
            }
        }
        return super.dispatchTouchEvent(ev);
    }

    @Override
    protected void onAttachedToWindow() {
        super.onAttachedToWindow();
        isAttachedToWindow = true;
        mapView.onStart();
        mapView.setId(getId());
        setupLayoutHack();
    }

    @Override
    protected void onDetachedFromWindow() {
        isAttachedToWindow = false;
        if (mapView != null)
            mapView.onSaveInstanceState(instanceStateBundle);
        super.onDetachedFromWindow();
    }

    public void onDropViewInstance() {
        mapView.onStop();
        mapView.onDestroy();
        removeAllViews();
        instanceStateBundle.clear();
        instanceStateBundle = null;
        mapView = null;
    }

    @Override
    public void onStop() {
        if (mapView != null)
            mapView.onStop();
    }

    @Override
    public void setCenter(LatLng latLng) {
        if (mapView != null)
            mapView.setCenter(latLng);
    }

    @Override
    public void setCenter(LatLng latLng, Double zoom, Double tilt, Double bearing) {
        if (mapView != null)
            mapView.setCenter(latLng, zoom, tilt, bearing);
    }

    @Override
    public void zoomTo(LatLngBounds latLngBounds, int paddingInPx) {
        if (mapView != null)
            mapView.zoomTo(latLngBounds, paddingInPx);
    }

    @Override
    public void setTilt(int tilt) {
        if (mapView != null)
            mapView.setTilt(tilt);
    }

    @Override
    public void setBearing(int bearing) {
        if (mapView != null)
            mapView.setBearing(bearing);
    }

    @Override
    public void setZoom(float zoom) {
        if (mapView != null)
            mapView.setZoom(zoom);
    }

    @Override
    public void setMapPadding(int left, int top, int right, int bottom) {
        if (mapView != null)
            mapView.setMapPadding(left, top, right, bottom);
    }

    @Override
    public void onInitialized() {
        if (mapView != null)
            mapView.onInitialized();
    }

    @Override
    public void showsMyLocationButton(boolean show) {
        if (mapView != null)
            mapView.showsMyLocationButton(show);
    }

    @Override
    public void setCompassEnabled(boolean show) {
        if (mapView != null)
            mapView.setCompassEnabled(show);
    }

    @Override
    public void setScaleBarEnabled(boolean show) {
        if (mapView != null)
            mapView.setScaleBarEnabled(show);
    }

    @Override
    public void setZoomControlEnabled(boolean show) {
        if (mapView != null)
            mapView.setZoomControlEnabled(show);
    }

    @Override
    public void setLocationTrackingMode(int mode) {
        if (mapView != null)
            mapView.setLocationTrackingMode(mode);
    }

    @Override
    public void setMapType(NaverMap.MapType value) {
        if (mapView != null)
            mapView.setMapType(value);
    }

    @Override
    public void setMinZoom(float minZoom) {
        if (mapView != null)
            mapView.setMinZoom(minZoom);
    }

    @Override
    public void setMaxZoom(float maxZoom) {
        if (mapView != null)
            mapView.setMaxZoom(maxZoom);
    }

    @Override
    public void setBuildingHeight(float height) {
        if (mapView != null)
            mapView.setBuildingHeight(height);
    }

    @Override
    public void setLayerGroupEnabled(String layerGroup, boolean enable) {
        if (mapView != null)
            mapView.setLayerGroupEnabled(layerGroup, enable);
    }

    @Override
    public void setNightModeEnabled(boolean enable) {
        if (mapView != null)
            mapView.setNightModeEnabled(enable);
    }

    @Override
    public void setLogoMargin(int left, int top, int right, int bottom) {
        if (mapView != null)
            mapView.setLogoMargin(left, top, right, bottom);
    }

    @Override
    public void setLogoGravity(int gravity) {
        if (mapView != null)
            mapView.setLogoGravity(gravity);
    }

    @Override
    public void setScrollGesturesEnabled(boolean enabled) {
        if (mapView != null)
            mapView.setScrollGesturesEnabled(enabled);
    }

    @Override
    public void setZoomGesturesEnabled(boolean enabled) {
        if (mapView != null)
            mapView.setZoomGesturesEnabled(enabled);
    }

    @Override
    public void setTiltGesturesEnabled(boolean enabled) {
        if (mapView != null)
            mapView.setTiltGesturesEnabled(enabled);
    }

    @Override
    public void setRotateGesturesEnabled(boolean enabled) {
        if (mapView != null)
            mapView.setRotateGesturesEnabled(enabled);
    }

    @Override
    public void setStopGesturesEnabled(boolean enabled) {
        if (mapView != null)
            mapView.setStopGesturesEnabled(enabled);
    }

    @Override
    public void setLiteModeEnabled(boolean enabled) {
        if (mapView != null)
            mapView.setLiteModeEnabled(enabled);
    }

    @Override
    public void moveCameraFitBound(LatLngBounds bounds, int left, int top, int right, int bottom) {
        if (mapView != null)
            mapView.moveCameraFitBound(bounds, left, top, right, bottom);
    }

    @Override
    public void addFeature(View child, int index) {
        if (mapView != null)
            mapView.addFeature(child, index);
    }

    @Override
    public void removeFeatureAt(int index) {
        if (mapView != null)
            mapView.removeFeatureAt(index);
    }

    @Override
    public int getFeatureCount() {
        if (mapView != null)
            return mapView.getFeatureCount();
        return 0;
    }

    @Override
    public View getFeatureAt(int index) {
        if (mapView != null)
            return mapView.getFeatureAt(index);
        return null;
    }
}

File: android/src/main/java/com/github/quadflask/react/navermap/RNNaverMapViewContainer.kt
package com.github.quadflask.react.navermap

import android.os.Bundle
import android.util.Log
import android.view.Choreographer
import android.view.MotionEvent
import android.view.View
import android.widget.FrameLayout
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.uimanager.ThemedReactContext
import com.github.quadflask.react.navermap.util.ReactUtil
import com.naver.maps.geometry.LatLng
import com.naver.maps.geometry.LatLngBounds
import com.naver.maps.map.NaverMap.MapType
import com.naver.maps.map.NaverMapOptions
import com.naver.maps.map.util.FusedLocationSource

class RNNaverMapViewContainer(
  themedReactContext: ThemedReactContext,
  appContext: ReactApplicationContext?,
  locationSource: FusedLocationSource?,
  naverMapOptions: NaverMapOptions?
) : FrameLayout(ReactUtil.getNonBuggyContext(themedReactContext, appContext!!)!!), RNNaverMapViewProps {
  private var mapView: RNNaverMapView?
  private var instanceStateBundle: Bundle? = Bundle()
  private var isAttachedToWindowFlag = false

  init {
    mapView = RNNaverMapView(themedReactContext, appContext, locationSource, naverMapOptions)
    addView(mapView)
    Log.e("RNN", "RNNaverMapViewContainer.init")
  }

  // https://github.com/facebook/react-native/issues/17968#issuecomment-457236577
  private fun setupLayoutHack() {
    Choreographer.getInstance().postFrameCallback(object : Choreographer.FrameCallback {
      override fun doFrame(frameTimeNanos: Long) {
        manuallyLayoutChildren()
        viewTreeObserver.dispatchOnGlobalLayout()
        if (isAttachedToWindowFlag) Choreographer.getInstance().postFrameCallbackDelayed(this, 500)
      }
    })
  }

  private fun manuallyLayoutChildren() {
    for (i in 0 until childCount) {
      val child = getChildAt(i)
      child.measure(
        MeasureSpec.makeMeasureSpec(measuredWidth, MeasureSpec.EXACTLY),
        MeasureSpec.makeMeasureSpec(measuredHeight, MeasureSpec.EXACTLY)
      )
      child.layout(0, 0, child.measuredWidth, child.measuredHeight)
    }
  }

  override fun dispatchTouchEvent(ev: MotionEvent): Boolean {
    mapView?.getMapAsync { map -> // TODO check async event dispatch
      if (map.uiSettings.isScrollGesturesEnabled) {
        when (ev.action) {
          MotionEvent.ACTION_DOWN,
          MotionEvent.ACTION_UP -> parent.requestDisallowInterceptTouchEvent(true)
        }
      }
    }
    return super.dispatchTouchEvent(ev)
  }

  override fun onAttachedToWindow() {
    super.onAttachedToWindow()
    isAttachedToWindowFlag = true
    mapView?.onCreate(instanceStateBundle)
    mapView?.onStart()
    mapView?.id = id
    setupLayoutHack()
    Log.e("RNN", "RNNaverMapViewContainer.onAttachedToWindow() - featureCount: ${mapView?.featureCount}")
  }

  override fun onDetachedFromWindow() {
    isAttachedToWindowFlag = false
    mapView?.onSaveInstanceState(instanceStateBundle!!)
    super.onDetachedFromWindow()
    Log.e("RNN", "RNNaverMapViewContainer.onDetachedFromWindow")
  }

  fun onDropViewInstance() {
    Log.e("RNN", "RNNaverMapViewContainer.onDropViewInstance - featureCount: ${mapView?.featureCount}")
    mapView?.onStop()
    mapView?.onDestroy()
    removeAllViews()
    instanceStateBundle!!.clear()
    instanceStateBundle = null
    mapView = null
  }

  override fun onStop() {
    Log.e("RNN", "RNNaverMapViewContainer.onStop")
    mapView?.onStop()
  }

  override fun setCenter(latLng: LatLng?) {
    mapView?.setCenter(latLng)
  }

  override fun setCenter(latLng: LatLng?, zoom: Double?, tilt: Double?, bearing: Double?) {
    mapView?.setCenter(latLng, zoom, tilt, bearing)
  }

  override fun zoomTo(latLngBounds: LatLngBounds, paddingInPx: Int) {
    mapView?.zoomTo(latLngBounds, paddingInPx)
  }

  override fun setTilt(tilt: Int) {
    mapView?.setTilt(tilt)
  }

  override fun setBearing(bearing: Int) {
    mapView?.setBearing(bearing)
  }

  override fun setZoom(zoom: Float) {
    mapView?.setZoom(zoom)
  }

  override fun setMapPadding(left: Int, top: Int, right: Int, bottom: Int) {
    mapView?.setMapPadding(left, top, right, bottom)
  }

  override fun onInitialized() {
    mapView?.onInitialized()
  }

  override fun showsMyLocationButton(show: Boolean) {
    mapView?.showsMyLocationButton(show)
  }

  override fun setCompassEnabled(show: Boolean) {
    mapView?.setCompassEnabled(show)
  }

  override fun setScaleBarEnabled(show: Boolean) {
    mapView?.setScaleBarEnabled(show)
  }

  override fun setZoomControlEnabled(show: Boolean) {
    mapView?.setZoomControlEnabled(show)
  }

  override fun setLocationTrackingMode(mode: Int) {
    mapView?.setLocationTrackingMode(mode)
  }

  override fun setMapType(value: MapType?) {
    mapView?.setMapType(value)
  }

  override fun setMinZoom(minZoomLevel: Float) {
    mapView?.setMinZoom(minZoomLevel)
  }

  override fun setMaxZoom(maxZoomLevel: Float) {
    mapView?.setMaxZoom(maxZoomLevel)
  }

  override fun setExtent(bounds: LatLngBounds?) {
    mapView?.setExtent(bounds)
  }

  override fun setBuildingHeight(height: Float) {
    mapView?.setBuildingHeight(height)
  }

  override fun setLayerGroupEnabled(layerGroup: String, enable: Boolean) {
    mapView?.setLayerGroupEnabled(layerGroup, enable)
  }

  override fun setNightModeEnabled(enable: Boolean) {
    mapView?.setNightModeEnabled(enable)
  }

  override fun setLogoMargin(left: Int, top: Int, right: Int, bottom: Int) {
    mapView?.setLogoMargin(left, top, right, bottom)
  }

  override fun setLogoGravity(gravity: Int) {
    mapView?.setLogoGravity(gravity)
  }

  override fun setScrollGesturesEnabled(enabled: Boolean) {
    mapView?.setScrollGesturesEnabled(enabled)
  }

  override fun setZoomGesturesEnabled(enabled: Boolean) {
    mapView?.setZoomGesturesEnabled(enabled)
  }

  override fun setTiltGesturesEnabled(enabled: Boolean) {
    mapView?.setTiltGesturesEnabled(enabled)
  }

  override fun setRotateGesturesEnabled(enabled: Boolean) {
    mapView?.setRotateGesturesEnabled(enabled)
  }

  override fun setStopGesturesEnabled(enabled: Boolean) {
    mapView?.setStopGesturesEnabled(enabled)
  }

  override fun setLiteModeEnabled(enabled: Boolean) {
    mapView?.setLiteModeEnabled(enabled)
  }

  override fun moveCameraFitBound(bounds: LatLngBounds, left: Int, top: Int, right: Int, bottom: Int) {
    mapView?.moveCameraFitBound(bounds, left, top, right, bottom)
  }

  override fun addFeature(child: View?, index: Int) {
    mapView?.addFeature(child, index)
    Log.e("RNN", "RNNaverMapViewContainer.addFeature(child: ${child?.javaClass?.simpleName}, index: $index)")
  }

  override fun removeFeatureAt(index: Int) {
    mapView?.removeFeatureAt(index)
    Log.e("RNN", "RNNaverMapViewContainer.removeFeatureAt(index: $index)")
  }

  override val featureCount: Int
    get() = mapView?.featureCount ?: 0

  override fun getFeatureAt(index: Int): View? {
    return mapView?.getFeatureAt(index)
  }
}
